### PR TITLE
Fix Elvis errors in ejabberd_c2s and introduce mongoose_transport

### PR DESCRIPTION
**Pull request by fenek**
Proposed changes include:
* All Elvis errors are fixed in `ejabberd_c2s` except for the one about module having more than 32 functions. It can't be resolved without more serious refactoring and breaking this file into several smaller ones.
* Added `mongoose_transport` behaviour. It's not final yet, because it will be improved when `ejabberd_receiver` and/or `ejabberd_listener` get rocked too.

I'm aware that `mongoose_transport` should take care of maintaining unified state with transport module name inside but it would involve touching even more modules. Again, this will get fixed while rocking other stuff.

### CODE REVIEW

File: apps/ejabberd/src/mod_websockets.erl [hunk 1]
```diff
@@ -248,6 +252,7 @@ get_sockmod(_SocketData) ->
 close(#websocket{pid = Pid}) ->
     Pid ! close.
 
+-spec peername(#websocket{}) -> mongoose_sockmod:peername_return().
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_websockets.erl]: According to **Elvis**:
> The spec in line 255 uses a record, please define a type for the record and use that instead.

File: apps/ejabberd/src/mod_websockets.erl [hunk 2]
```diff
@@ -224,6 +226,7 @@ reset_stream(#websocket{pid = Pid} = SocketData) ->
     Pid ! reset_stream,
     SocketData.
 
+-spec send_xml(#websocket{}, mongoose_sockmod:send_xml_input()) -> ok.
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_websockets.erl]: According to **Elvis**:
> The spec in line 229 uses a record, please define a type for the record and use that instead.

File: apps/ejabberd/src/mod_websockets.erl [hunk 3]
```diff
@@ -239,6 +242,7 @@ send(#websocket{pid = Pid}, Data) ->
 change_shaper(SocketData, _Shaper) ->
     SocketData. %% TODO: we ignore shapers for now
 
+-spec monitor(#websocket{}) -> reference().
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_websockets.erl]: According to **Elvis**:
> The spec in line 245 uses a record, please define a type for the record and use that instead.

File: apps/ejabberd/src/mod_bosh_socket.erl [hunk 4]
```diff
@@ -1,6 +1,7 @@
 -module(mod_bosh_socket).
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_bosh_socket.erl]: According to **Elvis**:
> This module has too many functions (34). Consider breaking it into a number of modules.

File: apps/ejabberd/src/ejabberd_socket.erl [hunk 5]
```diff
@@ -27,6 +27,8 @@
 -module(ejabberd_socket).
 -author('alexey@process-one.net').
 
+-behaviour(mongoose_sockmod).
+
 %% API
 -export([start/4,
 	 connect/3,
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_socket.erl]: According to **Elvis**:
> Line 34 has a tab at column 0.

**Comment by michalwski** [comment]
Thanks!

**Pull request merged**